From c6a1e59909c66665ed688825d13317e7fe1547d9 Mon Sep 17 00:00:00 2001
From: Martin Pitt <martinpitt@gnome.org>
Date: Tue, 12 Mar 2013 11:00:28 +0100
Subject: [PATCH] Add poweroff/reboot/suspend/hibernate fallback

When running logind without systemd, fall back to calling
poweroff/reboot/pm-suspend/pm-hibernate. This also preserves our current
pm-suspend quirks.
---
 src/login/logind-dbus.c | 45 +++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 43 insertions(+), 2 deletions(-)

Index: b/src/login/logind-dbus.c
===================================================================
--- a/src/login/logind-dbus.c
+++ b/src/login/logind-dbus.c
@@ -1043,6 +1043,33 @@
                           q, NULL);
 }
 
+/* Called if org.freedesktop.systemd1 is not available, i. e. running
+ * standalone logind */
+static int execute_shutdown_or_sleep_fallback(
+                Manager *m,
+                const char *unit_name) {
+    const char* cmd;
+
+    if (streq(unit_name, "suspend.target"))
+            cmd = "/usr/sbin/pm-suspend";
+    else if (streq(unit_name, "hibernate.target"))
+            cmd = "/usr/sbin/pm-hibernate";
+    else if (streq(unit_name, "reboot.target"))
+            cmd = "/sbin/reboot";
+    else if (streq(unit_name, "shutdown.target") ||
+             streq(unit_name, "poweroff.target"))
+            cmd = "/sbin/poweroff";
+    else {
+            log_error("execute_shutdown_or_sleep_fallback: unknown unit name %s", unit_name);
+            return 0;
+    }
+
+    log_info("execute_shutdown_or_sleep_fallback: Running command '%s' for unit %s", cmd, unit_name);
+    return system(cmd);
+}
+
+static int send_prepare_for(Manager *m, InhibitWhat w, bool _active);
+
 static int execute_shutdown_or_sleep(
                 Manager *m,
                 InhibitWhat w,
@@ -1072,8 +1099,23 @@
                         DBUS_TYPE_STRING, &unit_name,
                         DBUS_TYPE_STRING, &mode,
                         DBUS_TYPE_INVALID);
-        if (r < 0)
-                return r;
+        if (r < 0) {
+            log_info("execute_shutdown_or_sleep: failed to call org.freedesktop.systemd1, using fallback: %s", bus_error_message(error));
+            dbus_error_free(error);
+
+            r = execute_shutdown_or_sleep_fallback(m, unit_name);
+
+            /* In the fallback case, we're running standalone logind and so the
+             * 'false' events won't be sent for us by the job finishing. Send
+             * them explicitly here as the best thing we can do. */
+            send_prepare_for (m, w, false);
+
+            free(m->action_job);
+            m->action_job = NULL;
+            m->action_unit = NULL;
+            m->action_what = 0;
+            return r;
+        }
 
         if (!dbus_message_get_args(
                             reply,
